From 731980d76faac4e83768bc96caa2c67fb32d760c Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Wed, 6 Aug 2014 03:50:53 -0400
Subject: [pci] Check all IORESOURCE_TYPE_BITS in pci_bus_alloc_from_region()

Message-id: <20140806035052.1438.7092.stgit@gir.stowe>
Patchwork-id: 87472
O-Subject: [RHEL7.1 PATCH 63/69] PCI: Check all IORESOURCE_TYPE_BITS in pci_bus_alloc_from_region()
Bugzilla: 1110896
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110896
Upstream Status: aa11fc58dc71c27701b1f9a529a36a38d4337722

commit aa11fc58dc71c27701b1f9a529a36a38d4337722
Author: Bjorn Helgaas <bhelgaas@google.com>
Date:   Fri Mar 7 13:39:01 2014 -0700

    PCI: Check all IORESOURCE_TYPE_BITS in pci_bus_alloc_from_region()

    When allocating space from a bus resource, i.e., from apertures leading to
    this bus, make sure the entire resource type matches.  The previous code
    assumed the IORESOURCE_TYPE_BITS field was a bitmask with only a single bit
    set, but this is not true.  IORESOURCE_TYPE_BITS is really an enumeration,
    and we have to check all the bits.

    See 72dcb1197228 ("resources: Add register address resource type").

    No functional change.  If we used this path for allocating IRQs, DMA
    channels, or bus numbers, this would fix a bug because those types are
    indistinguishable when masked by IORESOURCE_IO | IORESOURCE_MEM.  But we
    don't, so this shouldn't make any difference.

    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/pci/bus.c b/drivers/pci/bus.c
index 3890166..e3bdc88 100644
--- a/drivers/pci/bus.c
+++ b/drivers/pci/bus.c
@@ -144,7 +144,7 @@ static int pci_bus_alloc_from_region(struct pci_bus *bus, struct resource *res,
  struct resource *r, avail;
  resource_size_t max;
 
- type_mask |= IORESOURCE_IO | IORESOURCE_MEM;
+ type_mask |= IORESOURCE_TYPE_BITS;
 
  pci_bus_for_each_resource(bus, r, i) {
   if (!r)
-- 
1.7.1